File: ios/ExpoDrawingModule.swift
import ExpoModulesCore

public class ExpoDrawingModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoDrawing')` in JavaScript.
    Name("ExpoDrawing")

    // Enables the module to be used as a native view. Definition components that are accepted as part of the
    // view definition: Prop, Events.
    View(ExpoDrawingView.self) {
      // Define events that can be sent to JavaScript
      Events("onStrokeStart", "onStrokeEnd")
      
      // MARK: - Props
      
      Prop("strokeColor") { (view: ExpoDrawingView, colorString: String?) in
        if let color = colorString {
          view.setStrokeColor(color)
        }
      }
      
      Prop("strokeThickness") { (view: ExpoDrawingView, thickness: Double?) in
        if let width = thickness {
          view.setStrokeThickness(CGFloat(width))
        }
      }
      
      Prop("tool") { (view: ExpoDrawingView, toolName: String?) in
        if let tool = toolName {
          view.setTool(tool)
        }
      }
      
      Prop("enableFingerDrawing") { (view: ExpoDrawingView, enabled: Bool?) in
        if let allow = enabled {
          view.setEnableFingerDrawing(allow)
        }
      }
      
      // MARK: - Methods
      
      AsyncFunction("undo") { (view: ExpoDrawingView) in
        view.undo()
      }
      
      AsyncFunction("redo") { (view: ExpoDrawingView) in
        view.redo()
      }
      
      AsyncFunction("clearDrawing") { (view: ExpoDrawingView) in
        view.clearDrawing()
      }
      
      AsyncFunction("getCanvasDataAsBase64") { (view: ExpoDrawingView) -> String? in
        return view.getCanvasDataAsBase64()
      }
      
      // iOS-only tool picker methods
      AsyncFunction("showToolPicker") { (view: ExpoDrawingView) in
        view.showToolPicker()
      }
      
      AsyncFunction("hideToolPicker") { (view: ExpoDrawingView) in
        view.hideToolPicker()
      }
    }
  }
}

File: android/src/main/java/expo/modules/drawing/ExpoDrawingModule.kt
package expo.modules.drawing

import android.graphics.Color
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoDrawingModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoDrawing")

    // Defines constant property on the module.
    Constant("PI") {
      Math.PI
    }

    // Defines event names that the module can send to JavaScript.
    Events("onChange")

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("hello") {
      "Hello world! 👋"
    }

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("setValueAsync") { value: String ->
      // Send an event to JavaScript.
      sendEvent("onChange", mapOf(
        "value" to value
      ))
    }

    // Drawing view component definition
    View(ExpoDrawingView::class) {
      // Events for stroke lifecycle
      Events("onStrokeStart", "onStrokeEnd")

      // Props for brush configuration
      Prop("strokeColor") { view: ExpoDrawingView, colorString: String? ->
        colorString?.let {
          try {
            val color = Color.parseColor(it)
            view.setStrokeColor(color)
          } catch (e: IllegalArgumentException) {
            // Invalid color string, ignore
          }
        }
      }

      Prop("strokeThickness") { view: ExpoDrawingView, thickness: Float? ->
        thickness?.let {
          if (it > 0) {
            view.setStrokeThickness(it)
          }
        }
      }

      Prop("tool") { view: ExpoDrawingView, toolName: String? ->
        toolName?.let {
          view.setTool(it)
        }
      }

      Prop("enableFingerDrawing") { view: ExpoDrawingView, allow: Boolean? ->
        allow?.let {
          view.setEnableFingerDrawing(it)
        }
      }

      // AsyncFunctions for imperative methods
      AsyncFunction("undo") { view: ExpoDrawingView ->
        view.undo()
      }

      AsyncFunction("redo") { view: ExpoDrawingView ->
        view.redo()
      }

      AsyncFunction("clearDrawing") { view: ExpoDrawingView ->
        view.clearDrawing()
      }

      AsyncFunction("getCanvasDataAsBase64") { view: ExpoDrawingView ->
        view.getCanvasDataAsBase64()
      }
    }
  }
}
